nucliadb/nucliadb/writer/back_pressure.py

```diff
@@ -0,0 +1,412 @@
+# Copyright (C) 2021 Bosutech XXI S.L.
+#
+# nucliadb is offered under the AGPL v3.0 and as commercial software.
+# For commercial licensing, contact us at info@nuclia.com.
+#
+# AGPL:
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+
+import asyncio
+import threading
+from datetime import datetime, timedelta
+from typing import Optional
+
+import nats
+from async_lru import alru_cache
+from cachetools import TTLCache
+from fastapi import HTTPException, Request
+from nucliadb_protos.writer_pb2 import ShardObject
+
+from nucliadb.common.context import ApplicationContext
+from nucliadb.common.context.fastapi import get_app_context
+from nucliadb.common.datamanagers.resources import ResourcesDataManager
+from nucliadb.common.http_clients.processing import ProcessingHTTPClient
+from nucliadb.writer import logger
+from nucliadb.writer.settings import back_pressure_settings as settings
+from nucliadb_telemetry import errors
+from nucliadb_utils import const
+from nucliadb_utils.nats import NatsConnectionManager
+from nucliadb_utils.settings import is_onprem_nucliadb
+
+"""
+TODO:
+- In the event of an unexpected error (tikv, nats, etc),
```

Comment:
it's really hard to say isn't it! Because maybe it's down because it's overloaded.